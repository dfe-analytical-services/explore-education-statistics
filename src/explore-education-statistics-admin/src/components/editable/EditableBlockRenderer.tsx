import EditableBlockWrapper from '@admin/components/editable/EditableBlockWrapper';
import EditableContentBlock from '@admin/components/editable/EditableContentBlock';
import { EditableBlock } from '@admin/services/publicationService';
import { GetInfographic } from '@common/modules/charts/components/InfographicBlock';
import DataBlockRenderer from '@common/modules/find-statistics/components/DataBlockRenderer';
import React, { useCallback, useMemo } from 'react';

interface Props {
  allowHeadings?: boolean;
  block: EditableBlock;
  editable?: boolean;
  getInfographic?: GetInfographic;
  onContentSave: (blockId: string, content: string) => void;
  onDelete: (blockId: string) => void;
}

function EditableBlockRenderer({
  allowHeadings,
  block,
  editable,
  getInfographic,
  onContentSave,
  onDelete,
}: Props) {
  const id = `editableBlockRenderer-${block.id}`;

  const handleContentSave = useMemo(
    () => (content: string) => {
      onContentSave(block.id, content);
    },
    [block.id, onContentSave],
  );

  const handleDelete = useCallback(() => {
    onDelete(block.id);
  }, [block.id, onDelete]);

  switch (block.type) {
    case 'DataBlock':
      return (
        <div className="dfe-content-overflow">
          <EditableBlockWrapper onDelete={editable ? handleDelete : undefined}>
            <DataBlockRenderer
              id={id}
              dataBlock={block}
              getInfographic={getInfographic}
            />
          </EditableBlockWrapper>
        </div>
      );
    case 'HtmlBlock':
    case 'MarkDownBlock':
      return (
        <EditableContentBlock
          allowHeadings={allowHeadings}
          editable={editable}
          id={id}
          label="Block content"
          value={block.body}
          useMarkdown={block.type === 'MarkDownBlock'}
          onSave={handleContentSave}
          onDelete={handleDelete}
        />
      );
    default:
      return <div>Unable to edit content</div>;
  }
}

export default EditableBlockRenderer;
