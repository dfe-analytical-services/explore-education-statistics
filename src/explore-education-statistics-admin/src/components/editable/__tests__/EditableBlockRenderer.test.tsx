import EditableBlockRenderer from '@admin/components/editable/EditableBlockRenderer';
import { render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import userEvent from '@testing-library/user-event';

describe('EditableBlockRenderer', () => {
  describe('MarkDownBlock', () => {
    test('renders non-editable version correctly', () => {
      const { container } = render(
        <EditableBlockRenderer
          releaseId="test-release-id"
          block={{
            id: 'block-000',
            order: 0,
            comments: [],
            type: 'MarkDownBlock',
            body: 'test',
          }}
          onContentSave={() => {}}
          onDelete={() => {}}
        />,
      );

      expect(screen.queryByText('Edit block')).toBeNull();
      expect(screen.queryByText('Remove block')).toBeNull();
      expect(container.innerHTML).toMatchSnapshot();
    });

    test('renders editable version correctly', () => {
      const { container } = render(
        <EditableBlockRenderer
          releaseId="test-release-id"
          editable
          block={{
            id: 'block-000',
            order: 0,
            comments: [],
            type: 'MarkDownBlock',
            body: 'test',
          }}
          onContentSave={() => {}}
          onDelete={() => {}}
        />,
      );

      expect(screen.queryByText('Edit block')).not.toBeNull();
      expect(screen.queryByText('Remove block')).not.toBeNull();
      expect(container.innerHTML).toMatchSnapshot();
    });

    test('clicking "Edit block" and "Save" calls save callback', () => {
      const handleContentSave = jest.fn();

      render(
        <EditableBlockRenderer
          releaseId="test-release-id"
          editable
          block={{
            id: 'block-000',
            order: 0,
            comments: [],
            type: 'MarkDownBlock',
            body: '',
          }}
          onContentSave={handleContentSave}
          onDelete={() => {}}
        />,
      );

      userEvent.click(screen.getByText('Edit block'));
      userEvent.click(screen.getByText('Save'));

      expect(handleContentSave).toHaveBeenCalledWith('block-000', '');
    });

    test('clicking "Remove block" and "Confirm" calls delete callback', async () => {
      const handleDelete = jest.fn();

      render(
        <EditableBlockRenderer
          releaseId="test-release-id"
          editable
          block={{
            id: 'block-000',
            order: 0,
            comments: [],
            type: 'MarkDownBlock',
            body: '',
          }}
          onContentSave={() => {}}
          onDelete={handleDelete}
        />,
      );

      userEvent.click(screen.getByText('Remove block'));

      await waitFor(() => {
        expect(screen.getByText('Confirm')).toBeInTheDocument();
      });

      userEvent.click(screen.getByText('Confirm'));

      expect(handleDelete).toHaveBeenCalledWith('block-000');
    });
  });
});
