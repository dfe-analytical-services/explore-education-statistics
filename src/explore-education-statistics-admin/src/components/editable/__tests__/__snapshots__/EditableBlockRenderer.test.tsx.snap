// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableBlockRenderer MarkDownBlock renders editable version correctly 1`] = `
<article>
  <div class="block">
    <div class="preview readOnly">
      <div class="editButton"
           role="button"
           tabindex="0"
      >
        <div>
          <p>
            test
          </p>
        </div>
      </div>
    </div>
  </div>
  <div class="buttons">
    <button aria-disabled="false"
            class="govuk-button marginRight govuk-button--secondary"
            type="button"
    >
      Edit block
    </button>
    <button aria-disabled="false"
            class="govuk-button marginRight govuk-button--warning"
            type="button"
    >
      Remove block
    </button>
  </div>
</article>
`;

exports[`EditableBlockRenderer MarkDownBlock renders non-editable version correctly 1`] = `
<article>
  <div class="block">
    <div class="preview readOnly">
      <div>
        <p>
          test
        </p>
      </div>
    </div>
  </div>
  <div class="buttons">
  </div>
</article>
`;
